src/faces/FacesTool.vala

```diff
@@ -0,0 +1,971 @@
+            event_box.leave_notify_event.connect(face_widget.on_leave_notify_event);
+
+            face_widgets_layout.pack_start(event_box, false);
+
+            if (buttons_text_separator == null) {
+                buttons_text_separator = new Gtk.HSeparator();
+                face_widgets_layout.pack_end(buttons_text_separator, false);
+            }
+
+            face_widgets_layout.show_all();
+        }
+
+        private void edit_face(Gtk.Button button) {
+            FaceWidget widget = (FaceWidget) button.get_parent();
+
+            face_edit_requested(widget.label.get_text());
+        }
+
+        private void delete_face(Gtk.Button button) {
+            FaceWidget widget = (FaceWidget) button.get_parent();
+
+            face_delete_requested(widget.label.get_text());
+
+            widget.get_parent().destroy();
+
+            if (face_widgets_layout.get_children().length() == 1) {
+                buttons_text_separator.destroy();
+                buttons_text_separator = null;
+            }
+        }
+
+        private void on_face_hidden() {
+            face_hidden();
+        }
+
+        private void on_cancel_detection() {
+            detection_canceled();
+        }
+    }
+
+    public class EditingFaceToolWindow : EditingTools.EditingToolWindow {
+        public signal bool key_pressed(Gdk.EventKey event);
+
+        public Gtk.Entry entry;
+
+        private Gtk.Box layout = null;
+
+        public EditingFaceToolWindow(Gtk.Window container) {
+            base(container);
+
+            entry = new Gtk.Entry();
+
+            layout = new Gtk.Box(Gtk.Orientation.HORIZONTAL, CONTROL_SPACING);
+            layout.add(entry);
+
+            add(layout);
+        }
+
+        public override bool key_press_event(Gdk.EventKey event) {
+            return key_pressed(event) || base.key_press_event(event);
+        }
+    }
+
+    private class FaceDetectionJob : BackgroundJob {
+        private Gee.Queue<string> faces = null;
+        private string image_path;
+        private string output;
+
+        public FaceDetectionJob(FacesToolWindow owner, string image_path,
+            CompletionCallback completion_callback, Cancellable cancellable,
+            CancellationCallback cancellation_callback) {
+            base(owner, completion_callback, cancellable, cancellation_callback);
+
+            this.image_path = image_path;
+        }
+
+        public override void execute() {
+            try {
+                string[] argv = {
+                    AppDirs.get_facedetect_bin().get_path(),
+                    "--cascade=" + AppDirs.get_haarcascade_file().get_path(),
+                    "--scale=1.2",
+                    image_path
+                };
+                Process.spawn_sync(null, argv, null, SpawnFlags.STDERR_TO_DEV_NULL, null, out output);
+
+            } catch (SpawnError e) {
+                stderr.printf("Error trying to spawn face detection program: %s\n", e.message);
```

Comment:
Make that a criticla() log message. Since this is a background job, save the error, then in the finished callback, check for the error and show it to the user with AppWindow.error()